### PR TITLE
Simulation for load balancing logic.

### PR DESCRIPTION
**Pull request by RomanDzhabarov**

### CODE REVIEW

File: test/common/upstream/load_balancer_simulation_test.cc [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        zone_hosts.push_back(newTestHost(cluster_, url, 1, zone));
+      }
+
+      ret.insert({zone, std::move(zone_hosts)});
+    }
+
+    return ret;
+  };
+
+  void setupRuntime() {
```
Comment (mattklein123) [inline test/common/upstream/load_balancer_simulation_test.cc]: do in constructor

File: test/common/upstream/load_balancer_simulation_test.cc [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+           std::vector<uint32_t> healthy_destination_cluster) {
+    setupRuntime();
+    stats_.upstream_zone_count_.set(all_destination_cluster.size());
+
+    std::unordered_map<std::string, std::vector<HostPtr>> healthy_map =
+        generateMap(healthy_destination_cluster);
+    std::unordered_map<std::string, std::vector<HostPtr>> all_map =
+        generateMap(all_destination_cluster);
+
+    std::vector<HostPtr> originatingHosts = generateList(originating_cluster);
```
Comment (mattklein123) [inline test/common/upstream/load_balancer_simulation_test.cc]: originatingHosts naming

File: test/common/upstream/load_balancer_simulation_test.cc [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                << std::endl;
+    }
+  }
+
+  HostPtr selectOriginatingHost(const std::vector<HostPtr>& hosts) {
+    // Originating cluster should have roughly the same per host request distribution.
+    return hosts[random_.random() % hosts.size()];
+  }
+
+  std::vector<HostPtr> generateList(const std::vector<uint32_t>& hosts) {
```
Comment (mattklein123) [inline test/common/upstream/load_balancer_simulation_test.cc]: better name and function comment, what does this do

File: test/common/upstream/load_balancer_simulation_test.cc (hunk 1 above)
Comment (RomanDzhabarov) [inline test/common/upstream/load_balancer_simulation_test.cc]: fixed

File: test/common/upstream/load_balancer_simulation_test.cc (hunk 2 above)
Comment (RomanDzhabarov) [inline test/common/upstream/load_balancer_simulation_test.cc]: fixed

File: test/common/upstream/load_balancer_simulation_test.cc (hunk 3 above)
Comment (RomanDzhabarov) [inline test/common/upstream/load_balancer_simulation_test.cc]: fixed.

File: test/common/upstream/load_balancer_simulation_test.cc (hunk 1 above)
Comment (RomanDzhabarov) [inline test/common/upstream/load_balancer_simulation_test.cc]: oh, ok, moving there

**Pull request merged**